core/domain/exp_domain.py

```diff
@@ -1463,7 +1463,8 @@ def validate(self, strict: bool = False) -> None:
                         for outcome in all_outcomes:
```

Comment:
There is [this comment](https://github.com/oppia/oppia/pull/15827#discussion_r945628981) which I can't seem to figure out how to reply to (that line is not changed in this PR but is still problematic?) -- so I'll reply to it here.

I understand that Any is used in one function, but on looking at that function I think we need a Union type for the possible values of solutions and we should use that type there. Otherwise, importing Any at the top gives a blanket exception for anyone to use Any in the entire file.

Reply:
The reason why the solution's correct_answer is marked as Any: The collection of values that 'correct_answer' can take is too complex. I mean we have List[Set[str]], List[List[str]], Dict[str, str], List[str], str, and int ( these are the mentioned ones in docstring and used in tests ).
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L453
Now, the problem is that if we use Union of all these types then we have to use assert, and in some places, we have to use cast also.

The case where we have to use a cast:
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L584
If we do assert here like this `assert isinstance(solution_dict['correct_answer'], list)` then MyPy rule out all those type from Union where List is not used. So, it will assume that  solution_dict['correct_answer'] can be of any type from (List[Set[str]], List[List[str]], List[str]) so to narrow the type further we have to use cast. 

The same behaviour can be observed here https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L590

In exp_domain Any type is used at only 1 place: which is at https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/exp_domain.py#L2091
Now if we see this function, it can also accept different types of values, like

here, solution's correct_answer https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/exp_domain.py#L2160
And in here we are passing ruleInput's values that have it's own sub set of types
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/exp_domain.py#L2185
https://github.com/oppia/oppia/blob/4ba04d87b7e8aea3953287ae183c8e7d1c0031b6/core/domain/state_domain.py#L50
So, if do not use Any here then we have to use assert and casts again.

I also agree with your concern that some can also use Any but i think for that we have linters.

Reply:
I think dropping the Any and using the assertions is actually appropriate here. Each of the examples you point to is conditioned by something relating to the interaction type, and the explanation for the assertion would be something like "For interactions of type XYZ, the solution is always formatted as a ABC.".

That feels to me like a tighter condition than using Any. In general I think the priority is (a) don't use Any, (b) don't use assertions. (a) is more important than (b).